File: design_pattern/creational_patterns/kotlin/FactoryMethod.kt
package design_pattern.creational_patterns.kotlin

abstract class Animal {
    abstract fun makeNoise()
}

class Dog : Animal() {
    override fun makeNoise() {
        println("Gau Gau")
    }
}

class Cat : Animal() {
    override fun makeNoise() {
        println("Meow Meow")
    }
}

class Chicken : Animal() {
    override fun makeNoise() {
        println("O o o o")
    }
}

enum class AnimalType {
    DOG, CAT, CHICKEN
}

object AnimalFactory {
    fun createAnimal(type: AnimalType): Animal {
        return when (type) {
            AnimalType.DOG -> Dog()
            AnimalType.CAT -> Cat()
            AnimalType.CHICKEN -> Chicken()
        }
    }
}

fun main() {
    AnimalFactory.createAnimal(AnimalType.CHICKEN).makeNoise()
}

File: design_pattern/creational_patterns/dart/factory_method.dart
abstract class Animal {
  static Animal createAnimal(AnimalType type) {
    switch (type) {
      case AnimalType.Dog:
        return Dog();
      case AnimalType.Cat:
        return Cat();
      case AnimalType.Chicken:
        return Chicken();
    }
  }

  void makeNoise();
}

enum AnimalType { Dog, Cat, Chicken }

class Dog extends Animal {
  @override
  void makeNoise() {
    print('Gau Gau');
  }
}

class Cat extends Animal {
  @override
  void makeNoise() {
    print('Meow Meow');
  }
}

class Chicken extends Animal {
  @override
  void makeNoise() {
    print('O o o o');
  }
}
